### PR TITLE
PulseQobj and Json Schema updates for pulse simulator

### LINKED ISSUE #2053: Pulse simulator QobjConfig
### What is the expected enhancement?
Pulse simulator configuration needs additional fields: `hamiltonian`, `dt`, `noise`, `ode_options`.
These values should be validated and added to `PulseQobjConfig`.

### PR DESCRIPTION
**Pull request by nkanazawa1989**
### Summary
This adds optional fields to `PulseQobjConfigSchema` for pulse simulator.

closes #2053 

### Details and comments
In current spec, these fields are specified as required regardless of backend (device/simulator)
```
"meas_level",
"pulse_library",
"memory_slot_size",
"meas_return",
"qubit_lo_freq",
"meas_lo_freq",
"rep_time"
```
This PR adds optional fields with validation
```
"hamiltonian",
"noise"
"dt"
"ode_options"
```

example of simulator configuration:
```
hamiltonian = {
    'h_str': [
      '2*pi*v0*O0',
      'X0||D0'
    ],
    'vars': {'v0': 5.1},
    'osc': {},
    'qub': {'0': 2}
}
noise = {
    'qubit': {
        '0': {
            'Sm': 0.006
        }
    },
    'oscillator': {
        'n_th': {},
        'coupling': {}
    }
}
ode_options = {
    'atol': 1e-18,
    'nsteps': 50000,
    'rhs_reuse': False,
    'rhs_filename': 'rhs'
}
dt = 1.0
```

### CODE REVIEW

**Comment by nonhermitian** [comment]
Why is `memory_slot_size` required?  Also currently QASM jobs have the option `memory_slots` which is just equal to the number of qubits on the device.  For Pulse we have `memory_slots` and `memory_slot_size`.  How does the relate to the QASM notation.  Are all the qasm jobs of `memory_slot_size = 1`?  And if so, then this does not play well with the `Results` class.  Indeed, formatting the memory in the results class automatically assumes memory slots sizes equal to the number of qubits, i.e . num slots = 1 slot size = num qubits.

**Comment by dcmckayibm** [comment]
`memory_slot_size` is only needed for meas_level 0 in open pulse (so doesn't apply really to qasm jobs). Otherwise the size should be 1 (ie. a slot fits a bit for meas_level 2 or a complex float for meas_level 1).

**Comment by nkanazawa1989** [comment]
Ok, I think we need to update the contains of `QobjConfig` to be reasonable both for the simulator and real backends. And I also need to create proper schemas for each field of simulator configuration. I changed this PR to WIP.

**Comment by nkanazawa1989** [comment]
The configurations for simulator is combined under `simulator_spec` field. This field is optional, but all fields necessary to be specified are set `required` within `simulator_spec`.

And below fields are removed from `required`:
```
"memory_slot_size",
"qubit_lo_freq",
"meas_lo_freq",
"rep_time"
```

Current `PulseQobjConfig` comprises:
```
+-meas_level
+-meas_return
+-pulse_library
+-(memory_slot_size)
+-(qubit_lo_freq)
+-(meas_lo_freq)
+-(rep_time)
+-(simulator-spec)
    |
    +-hamiltonian
    +-vars
    +-dim_osc
    +-dim_qub
    +-dt
    +-(noise_model)
    +-(ode_options)
```

and `simulator_spec` can be crated by new models:
```
simulator_spec = PulseSimulatorSpec(
    hamiltonian=['2*pi*v0*O0', 'X0||D0'],
    vars={'v0': 5.1},
    dim_osc=[],
    dim_qub=[2],
    dt=1.0,
    noise_model=PulseNoiseModel(
        qubit=[{'Sm': 0.006}],
        oscillator=[]
    ),
    ode_options=PulseOdeOption(
        atol=1e-18,
        nsteps=50000,
        rhs_reuse=False,
        rhs_filename='rhs'
    )
)
```

**Comment by nkanazawa1989** [comment]
Finally the `PulseQobjConfig` becomes:
```
+-meas_level
+-meas_return
+-pulse_library
+-(memory_slot_size)
+-qubit_lo_freq
+-meas_lo_freq
+-(rep_time)
+-(simulator-spec)
```
where, optionals are in ().

`simulator-spec` will be defined for each pulse simulators.
For now, it comprises:
```
simulator-spec
    |
    +-hamiltonian
    |    +-h_str
    |    +-vars
    |    +-dim_osc
    |    +-dim_qub
    +-dt
    +-(noise_model)
    |    +-qubit
    |    +-oscillator
    +-(ode_options)
```
This field can be generated with validation by:
```
simulator_spec = PulseSimulatorSpec(
    hamiltonian=PulseHamiltonian(
        h_str=['2*pi*v0*O0', 'X0||D0'],
        vars={'v0': 5.1},
        dim_osc=[],
        dim_qub=[2]
    )
    dt=1.0,
    noise_model=PulseNoiseModel(
        qubit=[{'Sm': 0.006}],
        oscillator=[]
    ),
    ode_options=PulseOdeOption(
        atol=1e-18,
        nsteps=50000,
        rhs_reuse=False,
        rhs_filename='rhs'
    )
)
```

Also I allowed for empty list of `gates` and `basis_gates` in `backend_configuration_schema` to create pulse-simulator dedicated backend.

File: test/python/qobj/test_qobj.py [hunk 1]
```diff
@@ -148,7 +149,26 @@ def setUp(self):
                                    ],
                                    qubit_lo_freq=[4.9],
                                    meas_lo_freq=[6.9],
-                                   rep_time=1000),
+                                   rep_time=1000,
```
Comment (taalexander) [inline test/python/qobj/test_qobj.py]: By initializing the valid_qobj in the `setUp` method you are not actually running the test in the test. This should be a dictionary to start, or moved to the actual test.

File: qiskit/qobj/models/pulse.py [hunk 2]
```diff
@@ -10,20 +10,21 @@
...
+from .pulse_simulator import PulseSimulatorSpecSchema
 
 
 class QobjMeasurementOptionSchema(BaseSchema):
     """Schema for QobjMeasOptiton."""
 
     # Required properties.
     name = String(required=True)
-    params = MeasurementParameter(required=True)
+    params = DictParameters(valid_value_types=(int, float, str, bool, type(None)),
```
Comment (taalexander) [inline qiskit/qobj/models/pulse.py]: Does this still allow arbitrary nesting of dicts?

File: qiskit/qobj/models/pulse.py [hunk 3]
```diff
@@ -76,18 +77,21 @@ class PulseQobjExperimentSchema(QobjExperimentSchema):
...
-    qubit_lo_freq = List(Number(), required=True)
-    meas_lo_freq = List(Number(), required=True)
-    rep_time = Integer(required=True)
-    meas_return = String(validate=OneOf(choices=(MeasReturnType.AVERAGE,
-                                                 MeasReturnType.SINGLE)))
+    meas_lo_freq = List(Number(validate=Range(min=0)), required=True)
+    meas_return = String(required=True, validate=OneOf(choices=(MeasReturnType.AVERAGE,
+                                                                MeasReturnType.SINGLE)))
+    pulse_library = Nested(QobjPulseLibrarySchema, required=True, many=True)
+    qubit_lo_freq = List(Number(validate=Range(min=0)), required=True)
```
Comment (taalexander) [inline qiskit/qobj/models/pulse.py]: should be with meas_lo_freq for clarity.

File: qiskit/qobj/models/pulse.py [hunk 4]
```diff
@@ -182,29 +186,22 @@ class PulseQobjConfig(QobjConfig):
...
         self.qubit_lo_freq = qubit_lo_freq
-        self.meas_lo_freq = meas_lo_freq
-        self.rep_time = rep_time
 
         super().__init__(meas_level=meas_level,
-                         memory_slot_size=memory_slot_size,
+                         meas_lo_freq=meas_lo_freq,
                          meas_return=meas_return,
                          pulse_library=pulse_library,
                          qubit_lo_freq=qubit_lo_freq,
```
Comment (taalexander) [inline qiskit/qobj/models/pulse.py]: Align this as well.

File: qiskit/qobj/models/pulse_simulator.py [hunk 5]
```diff
@@ -0,0 +1,149 @@
...
+        required=True)
+    oscillator = List(
+        List(Number, validate=Length(equal=2)),
+        required=True
+    )
+
+    # Optional properties.
+
+
+class PulseOdeOptionSchema(BaseSchema):
```
Comment (taalexander) [inline qiskit/qobj/models/pulse_simulator.py]: I do not think this should live in the `terra` repository. These are specific to the `Aer` simulator and could differ from simulator to simulator. The model should live in the Aer repository and be used to validate the submitted parameters.

File: qiskit/schemas/qobj_schema.json [hunk 6]
```diff
@@ -999,16 +999,141 @@
...
+                                            },
+                                            "type": "array"
+                                        }
+                                    },
+                                    "required": [
+                                        "qubit",
+                                        "oscillator"
+                                    ]
+                                },
+                                "ode_options": {
```
Comment (taalexander) [inline qiskit/schemas/qobj_schema.json]: These should also be removed. Aer can always provide its own jsonschema with specific `ode_options` if required.

File: qiskit/qobj/models/pulse.py (hunk 2 above)
Comment (taalexander) [reply qiskit/qobj/models/pulse.py]: Yes it does.

**Pull request merged**